### ISSUE TITLE
[docdb] Consider batching DeleteTablet reports as part of acting on TS heartbeats

### ISSUE THREAD

**Issue by bmatican** [yugabyte/yugabyte-db#2716]
Jira Link: [DB-2506](https://yugabyte.atlassian.net/browse/DB-2506)
If a TS goes down for >5min, it will be kicked out of all raft groups for which it had tablets for.
When it wakes up and heartbeats to the master, each of these tablets will be deemed invalid and the master will send `DeleteTablet` RPCs for all of them. This is done sequentially now! The TS has a limited thread pool and queue (see #2151), so it will obviously not be able to process large amounts of these RPCs received in quick succession.

The master however sends these RPCs as part of background tasks, with retries and exponential backoff, so technically, the deletes will eventually get processed...However, this is needless network transfer and cpu/memory churn for both servers.

Filing this for later. Should evaluate if each individual call is "fast" or "slow", else batching could actually make things worse.

**Comment by joe-maley** [comment]
For my own understanding and because this ticket is almost 3 years old, I reproduced this scenario to observe the current behavior. I observed that there are two paths that trigger the DeleteTablet RPC retry loop:
1. In that path that handles `FLAGS_master_tombstone_evicted_tablet_replicas`, the heartbeats from the remaining tservers will launch an `AsyncDeleteReplica` task for each replica on the dead tserver. It appears that the retry limit is `INT_MAX`. 
2. When the dead tserver is rebooted, it will send a heartbeat to the master. The master detects that this tserver is no longer in the raft group and will issue `AsyncDeleteReplica` tasks for each replica.

Note: I don't know if this a legitimate issue or just related to my test environment, but I see that the first path is triggered every `FLAG_follower_unavailable_considered_failed_sec` interval that the tserver is down. This creates an unbounded number of retrying `AsyncDeleteReplica` tasks. This eventually crashes my master server.

I think both paths would benefit from both batching and throttling the requests. For this ticket, I'll focus on batching the requests. I'll either modify the existing `DeleteTabletRequestPB` or introduce a new `BatchedDeleteTabletRequestPB`. The batch size should linearly decrease the network churn, so even a small batch size of 5 would give us an 80% reduction in network churn.

**Comment by bmatican** [comment]
> Note: I don't know if this a legitimate issue or just related to my test environment, but I see that the first path is triggered every FLAG_follower_unavailable_considered_failed_sec interval that the tserver is down. This creates an unbounded number of retrying AsyncDeleteReplica tasks. This eventually crashes my master server.
@joe-maley Oh, that's strange! Can you elaborate on this? So you're saying we then build up DeleteTablet requests on the master, that each have an infinite timeout?

> I think both paths would benefit from both batching and throttling the requests. For this ticket, I'll focus on batching the requests. I'll either modify the existing DeleteTabletRequestPB or introduce a new BatchedDeleteTabletRequestPB. The batch size should linearly decrease the network churn, so even a small batch size of 5 would give us an 80% reduction in network churn.
An alternative, or maybe parallel idea, would be to expand on some of @sanketkedia 's work, for throttling Snapshot RPCs and have some generic component which throttles RPCs to a destination tserver. For example, in this case, we could say, no more than N outstanding DeleteTablet RPCs to a tserver, should be started. Moreover, if one is already in flight, don't create another one, as this operation is terminal.

The challenge with a net new RPC, is that you'd have to coordinate cluster upgrades, ie: don't use the new RPC, unless you know the tserver has the new code to accept the RPC. We have some infra for this (capabilities) and long term it might even be desirable, but with the power of hindsight, I am trying to reason through pros and cons..

**Comment by joe-maley** [comment]
> @joe-maley Oh, that's strange! Can you elaborate on this? So you're saying we then build up DeleteTablet requests on the master, that each have an infinite timeout?

Yes, that's what I see. I'll explain this in detail and let you judge the severity of the issue. My simplified test code looks like this:
```
// using --enable_leader_failure_detection=false
// using --follower_unavailable_considered_failed_sec=5
// using --catalog_manager_wait_for_new_tablets_to_elect_leader=false
// using --use_create_table_leader_hint=false

TestWorkload(cluster_.get()).Setup();

StartElection(leader_ts, tablet_id, ...);
WaitForServersToAgree(..., tablet_id, ...);

cluster_->tablet_server(kFollowerIndex)->Shutdown();
SleepFor(MonoDelta::FromSeconds(30));
```

Every ~5 seconds I see the following statement reported from the raft leader, where the first log statement has `commit index -1` and the last one has `commit index 9`.
```
[ts-1] ... raft_consensus.cc:1474] ... Attempting to remove follower ... from the Raft config at commit index -1. Reason: Leader has been unable to successfully communicate with Peer ... for more than 5 seconds (5.135s)
```

Each one of these log statements is followed by the master creating an `AsyncDeleteReplica` task in the following path:
```
bool CatalogManager::ProcessCommittedConsensusState(...) {
  ...
    if (FLAGS_master_tombstone_evicted_tablet_replicas) {
      ...
      rpcs->push_back(std::make_shared<AsyncDeleteReplica>(...));
    }
  ...
}
```

Each task retries up to `2147483647` times:
```
[m-1] ... async_rpc_tasks.cc:377] DeleteTablet RPC for tablet ... (task=0x0000000153114528, state=kRunning): Scheduling retry with a delay of 507ms (attempt = 38 / 2147483647)...
```

If I modify the delay to a small fixed interval, my master process crashes in about a minute.

**Comment by joe-maley** [comment]
> An alternative, or maybe parallel idea, would be to expand on some of @sanketkedia 's work, for throttling Snapshot RPCs and have some generic component which throttles RPCs to a destination tserver. For example, in this case, we could say, no more than N outstanding DeleteTablet RPCs to a tserver, should be started. Moreover, if one is already in flight, don't create another one, as this operation is terminal.

I did ask @sanketkedia about this yesterday and he shared his patch. I reasoned that both throttling and batching would be beneficial, but I thought that batching would have a greater impact and that both changes would take about the same amount of work and risk (more on this, next).

> The challenge with a net new RPC, is that you'd have to coordinate cluster upgrades, ie: don't use the new RPC, unless you know the tserver has the new code to accept the RPC. We have some infra for this (capabilities) and long term it might even be desirable, but with the power of hindsight, I am trying to reason through pros and cons..

Good point. I didn't consider this. If this was a high priority, I can envision an immediate work-around/hack by modifying the existing DeleteReplicaPB to include a field like `repeated bytes additional_tablet_ids` that would be ignored by old deployments, allowing the new clients to detect only one tablet was deleted and retry with N-1 tablet ids. That's an ugly hack so I'll go the throttling route for now so that I can do _something_ for this ticket.

**Comment by sanketkedia** [comment]
>Note: I don't know if this a legitimate issue or just related to my test environment, but I see that the first path is triggered every FLAG_follower_unavailable_considered_failed_sec interval that the tserver is down.

@joe-maley for this, I see in the code that we do check if the destination tserver is registered with the master and not removed, so technically after `tserver_unresponsive_timeout_ms` (default is 60 secs), we should stop queuing more requests from the HB path on the master. Is there a bug here with this logic?

```
if (FLAGS_master_tombstone_evicted_tablet_replicas) {
      std::unordered_set<string> current_member_uuids;
      for (const consensus::RaftPeerPB &peer : cstate.config().peers()) {
        InsertOrDie(&current_member_uuids, peer.permanent_uuid());
      }
      for (const consensus::RaftPeerPB &prev_peer : prev_cstate.config().peers()) {
        const string& peer_uuid = prev_peer.permanent_uuid();
        if (!ContainsKey(current_member_uuids, peer_uuid)) {
          // Don't delete a tablet server that hasn't reported in yet (Bootstrapping).
          shared_ptr<TSDescriptor> dummy_ts_desc;
          if (!master_->ts_manager()->LookupTSByUUID(peer_uuid, &dummy_ts_desc)) {
            continue;
          }
          // Otherwise, the TabletServer needs to remove this peer.
          rpcs->push_back(std::make_shared<AsyncDeleteReplica>(
              master_, AsyncTaskPool(), peer_uuid, tablet->table(), tablet->tablet_id(),
              TABLET_DATA_TOMBSTONED, prev_cstate.config().opid_index(),
              Substitute("TS $0 not found in new config with opid_index $1",
                  peer_uuid, cstate.config().opid_index())));
        }
      }
    }
```

**Comment by joe-maley** [comment]
Fixed by dd9bb0f00dd86198852866f2e73f45bc1fc902a0
